osquery/logger/plugins/aws_log_forwarder.h

```diff
@@ -0,0 +1,301 @@
+          send_error = false;
+          break;
+        }
+
+        if (is_last_retry) {
+          if (!status_output.str().empty()) {
+            status_output << "\n";
+          }
+
+          status_output << outcome.GetError().GetMessage();
+          error_count++;
+        }
+
+        // We didn't manage to send all records; remove the ones that succeeded
+        // (so that we do not duplicate them) and try again
+        const auto& result_record_list = getResult(outcome);
+
+        for (std::size_t i = batch.size(); i-- > 0;) {
+          if (result_record_list[i].GetErrorCode().empty()) {
+            auto it = std::next(batch.begin(), i);
+            batch.erase(it);
+          }
+        }
+      }
+
+      // We couldn't write some of the records; log them locally so that the
+      // administrator will at least be able to inspect them
+      if (send_error) {
+        std::stringstream error_output;
+        error_output << name_
+                     << " logger: Failed to write the following records:\n";
+        dumpBatch(error_output, batch);
+
+        LOG(ERROR) << error_output.str();
+      }
+
+      batch_it = batch_list.erase(batch_it);
+    }
+
+    if (error_count != 0) {
+      return Status(1, status_output.str());
+    }
+
+    return Status(0, "OK");
+  }
+
+  /// Plugin-specific initialization is performed here
+  virtual Status internalSetup() = 0;
+
+  /// Plugin-specific send method
+  virtual Outcome internalSend(const Batch& batch) = 0;
+
+  /// Plugin-specific record initialization is performed here
+  virtual void initializeRecord(Record& record,
+                                Aws::Utils::ByteBuffer& buffer) const = 0;
+
+  /// Must return the amount of bytes that can fit in a single record
+  virtual std::size_t getMaxBytesPerRecord() const = 0;
+
+  /// Must return the amount of records that can be inserted into a single batch
+  virtual std::size_t getMaxRecordsPerBatch() const = 0;
+
+  /// Must return the amount of bytes that can fit in a single batch
+  virtual std::size_t getMaxBytesPerBatch() const = 0;
+
+  /// Must return the maximum amount of retries when sending records
+  virtual std::size_t getMaxRetryCount() const = 0;
+
+  /// Must return the initial delay, in seconds, between each retry
+  virtual std::size_t getInitialRetryDelay() const = 0;
+
+  /// Must return true if records should be terminated with newlines
+  virtual bool appendNewlineSeparators() const = 0;
+
+  /// Must return the amount of records that could not be sent
+  virtual std::size_t getFailedRecordCount(Outcome& outcome) const = 0;
+
+  /// Must return the vector containing the upload result for each record
+  virtual Result getResult(Outcome& outcome) const = 0;
+
+ protected:
+  /// Plugin-specific service client
+  std::shared_ptr<Client> client_{nullptr};
+
+  /// Logger name; used when printing messages
+  std::string name_;
+};
+
+using IFirehoseLogForwarder = AwsLogForwarder<
```

Comment:
Could this header be 'lighter' if these were in the implementation files, and move the respective firehose/kinesis includes into those files too?

Reply:
Done